### ISSUE TITLE
brisk-reconciler.ppx prevents usage of other standard JSX apis

### ISSUE THREAD

**Issue by kyldvs** [briskml/brisk-reconciler#56]
I've noticed that the brisk-reconciler.ppx prevents using other JSX apis. I believe the ppx expects there to be a `make` function but the standard JSX api built in to reason uses a `createElement` function.

This comes up when trying to use something like [Pastel](https://reason-native.com/docs/pastel/).

```reason
print_endline(<Pastel color=Red> "Hello world" </Pastel>);
```

It also seems like the brisk reconciler ppx doesn't support things like `<View> ...someListOfElements </View>` like normal JSX.

**Comment by wokalski** [comment]
Good point... There are no technical limitations. In fact, instead of:

```
let%componet make = ...
```

People could type:

```
let%component createElement = ...
```

I like `make` because it mirrors reason-react but if the people wouldn't mind `createElement`, I wouldn't mind it either. The issue with `children` is much more serious. Special handling of children allows for:

1. Optional children
2. Explicitly not accepting children
3. Non-optional children
4. Special handling of JSX lists so that they are reconciled differently than dynamic lists.

The design is quite limiting. I wish there was a way to have different transforms. One (crude) solution would be to add `[@brisk.nojsx]` attribute that you can attach to arbitrary expression and the brisk jsx ppx won't be applied inside it.

**Comment by kyldvs** [comment]
I would strongly prefer `let%component createElement = ...` so that we don't restrict the set of features Reason offers when using this ppx.

---

Regarding children, I think 1-3 all have passable solutions with the standard treatment of children as a list.

1. I think this is the default, you can optionally pass or not pass children, (or if I misunderstand it can probably be solved the same as (3)).
2. When I want to design a component like this I mark the children as unit: `~children: list(unit)`. Now it is very hard to pass anything meaningful as children.
3. Here I would just use a prop rather than accepting the required element as a child. Not as nice syntactically I guess, but I prefer the slightly worse syntax over messing with the semantics of Reason's built in features.
4. I don't understand this enough to offer an alternative/determine importance, so maybe this one would warrant diverging :)

**Comment by wokalski** [comment]
1. You can. But it's restricted `~children=[]` is passed by default. You cannot control the default value for instance.
2. `~children: list(unit)` before we had any PPX `~children as _: list(unit)` was omnipresent. It felt good to get rid of it.
3. I disagree it's messing with the semantics of Reason's built in features. `[@JSX]` is there for a reason and `Reason-React` has been leveraging it for a long time. I'd say that if anything, the builtin Reason's handling of children is suboptimal.

The API currently used by brisk-reconciler is the most natural one to use. Especially with things like optional "render props". It just works intuitively.

**Comment by kyldvs** [comment]
Yeah I do really like the API we have with this PPX, and it is natural to use! The unfortunate aspect is that it splits the Reason language. We can either use "Brisk JSX" or "Reason JSX". I don't think these conveniences warrant that split, but I understand other may disagree, which is fine.

My ideal scenario would be to bring these improvements to Reason's JSX where possible, but that is probably a lot of work.

In the meantime do you have any ideas how we could let the two kinds of JSX live side-by-side? `[@brisk.nojsx]` seems like a good place to start, not sure if file-level or expression level is better though

**Comment by wokalski** [comment]
`[@brisk.nojsx]` could work at any level. The mapper is defined [here](https://github.com/briskml/brisk-reconciler/blob/master/jsx/brisk_jsx.ml#L59-L78). You could use [`Ast_traverse.map_with_context`](https://github.com/ocaml-ppx/ppxlib/blob/master/src/ast_traverse.mli#L50-L53) objects to implement a mapping which contains some additional context. The context would contain information whether the attribute in a parent expression/structure.

**Comment by wokalski** [comment]
@kyldvs I asked Revery users for feedback regarding the JSX ppx. We'll see what they say. The patch to change JSX in refmt would be very simple. I once implemented [a related change](https://github.com/facebook/reason/pull/1650) quickly. What's more @IwanKaramazow and Antonia are very helpful with those changes.

**Comment by glennsl** [comment]
A scoped `@brisk.nojsx` might be nice. Something like this perhaps:

```reason
[@brisk.nojsx
  Pastel.(print_endline(<Pastel color=Red> "Hello world" </Pastel>))
];
```

Mixing several different kinds of JSX worries me, so it would be nice if "external" JSX code was clearly marked.
